core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -399,4 +410,64 @@ private static boolean isSplitLocal(List<HostAddress> splitAddresses, HostAddres
         }
         return false;
     }
+
+    private static class QueueSizeAdjuster
+    {
+        private final Map<String, Long> splitsWeightPerNode;
+        private final Set<String> previousScheduleFullNodes;
+        private final Map<String, Long> previouslyScheduledWeight;
+        private final long minPendingSplitsWeightPerTask;
+        private final long maxPendingSplitsWeightPerTask;
+        private final long maxSplitsWeightPerNode;
+
+        QueueSizeAdjuster(long minPendingSplitsWeightPerTask, long maxPendingSplitsWeightPerTask, long maxSplitsWeightPerNode)
+        {
+            this.splitsWeightPerNode = new HashMap<>();
+            this.previousScheduleFullNodes = new HashSet<>();
+            this.previouslyScheduledWeight = new HashMap();
+            this.maxPendingSplitsWeightPerTask = maxPendingSplitsWeightPerTask;
+            this.minPendingSplitsWeightPerTask = minPendingSplitsWeightPerTask;
+            this.maxSplitsWeightPerNode = maxSplitsWeightPerNode;
+        }
+
+        public void update(List<RemoteTask> existingTasks, NodeAssignmentStats nodeAssignmentStats)
+        {
+            if (maxPendingSplitsWeightPerTask == minPendingSplitsWeightPerTask) {
+                return;
+            }
+            for (RemoteTask task : existingTasks) {
+                String nodeId = task.getNodeId();
+                long adjustedMaxSplitsWeightPerTask = getAdjustedMaxPendingSplitsWeightPerTask(nodeId);
+
+                if (adjustedMaxSplitsWeightPerTask < maxPendingSplitsWeightPerTask
+                        && previousScheduleFullNodes.contains(nodeId)
+                        && nodeAssignmentStats.getQueuedSplitsWeightForStage(nodeId) == 0) {
+                    adjustedMaxSplitsWeightPerTask = Math.min(maxPendingSplitsWeightPerTask, adjustedMaxSplitsWeightPerTask * 2);
+                }
+                else if (adjustedMaxSplitsWeightPerTask > minPendingSplitsWeightPerTask
+                        && !previousScheduleFullNodes.contains(nodeId)
+                        && (previouslyScheduledWeight.getOrDefault(nodeId, 0L) - nodeAssignmentStats.getQueuedSplitsWeightForStage(nodeId)) < (getMaxSplitsWeightPerNode(nodeId) / 2)) {
+                    adjustedMaxSplitsWeightPerTask = (long) Math.max(minPendingSplitsWeightPerTask, adjustedMaxSplitsWeightPerTask / 1.5);
+                }
+                previouslyScheduledWeight.put(nodeId, nodeAssignmentStats.getQueuedSplitsWeightForStage(nodeId));
+                splitsWeightPerNode.put(nodeId, adjustedMaxSplitsWeightPerTask);
```

Comment:
nit: what do you think about instead of having two maps with longs per node, have one map with  both values (new class to keep that would be needed)?

Reply:
in any case, would be good to add documentation to this mehod